### PR TITLE
Use application tsconfig file for indexTransform and customWebpackConfig

### PR DESCRIPTION
**Pull request by michelherv**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
The application's `tsconfig` is ignore. Accordingly, `indexTransform` and `customWebpackConfig` cannot use project configuration for custom modules for example

Issue Number: #607 #749 

## What is the new behavior?
The application's `tsconfig` is now use. `indexTransform` and `customWebpackConfig` can now use all the application resources.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by just-jeb** [comment]
Hi. First of all thanks for your contribution - it is highly appreciated!  
Your PR in some way intersects with #824, and the reason the build is failing is the same reason.  
If you have time and desire to investigate this issue, it will be incredibly helpful.

**Comment by just-jeb** [comment]
Hey, thanks for the contribution.  
Currently two things are missing here - updating e2e tests and docs.  
I'd start with e2e, since I have a strong suspicion they will start failing once you try to use `.ts` config with Karma builder like [here](https://github.com/just-jeb/angular-builders/pull/824/files#diff-a2d81be02f314726924854ce8ad329d2be1bb7f6d8fcc44f9cd9ad4d3c3f5963R103).

File: scripts/run-ci.sh [hunk 1]
```diff
@@ -46,6 +46,10 @@ publishToLocalRegistry
 
 yarn bootstrap:examples
 
+# Get travis's chrome version and download the appropriate webdriver-manager for protractor
+CHROME_VERSION=`google-chrome --version | egrep -o '[0-9.]+' | head -1`
+npx lerna exec --ignore '@angular-builders/*' -- ./node_modules/protractor/bin/webdriver-manager update --versions.chrome $CHROME_VERSION
```
Comment (muuvmuuv) [inline scripts/run-ci.sh:51]: What is the reason for this?

File: packages/custom-webpack/package.json [hunk 2]
```diff
@@ -42,6 +42,7 @@
     "@angular-devkit/core": "^11.0.0",
     "lodash": "^4.17.15",
     "ts-node": "^9.0.0",
+    "tsconfig-paths": "^3.9.0",
```
Comment (muuvmuuv) [inline packages/custom-webpack/package.json:45]: I am currently merging part of your changes into my PR (#824) and adding `tsconfig-paths` was not needed. Why are you adding it, are there any benefits?

Comment (just-jeb) [reply packages/custom-webpack/package.json:45]: As far as I remember `ts-node` does not support `paths` property, so if you want to have the exact same config (including paths) for your webpack (or index) configuration you need this one.

Comment (muuvmuuv) [reply packages/custom-webpack/package.json:45]: Makes sense! But `ts-node` isn't needed for any Angular project since it compiles through `tsc` and does not require to be "started" through Node, so I wouldn't add this or do I miss any usecase here?

Comment (just-jeb) [reply packages/custom-webpack/package.json:45]: `ts-node` is indeed not needed for Angular app, because Angular compile (not exactly `tsc`) builds it, that's correct.  
However, index transform and webpack config are not part of the app, they are part of the build chain, which in our case is `custom-webpack` builder. The builders needs `ts-node` in order to understand the TS syntax of the files it loads (webpack config and index transform).

File: scripts/run-ci.sh (hunk 1 above)
Comment (michelherv) [reply scripts/run-ci.sh]: I cherry picked this from #886. It was to skip resolved errors. Now that #886 is merged, I rebase this pull request onto master and remove cherry picked commits
Thanks for the question

**Comment by michelherv** [comment]
Hi,
I updated e2e tests. Can you read the pull request again and tell me if tests are ok for you. Else can you help me by pushing some tests? I will try to improve the solution by TDD.

About the documentation, I have no improvement idear. `tsConfig` is a `@angular-devkit/architect:browser` property. It's not like `customWebpackConfig` or `indexTransform` which are custom properties. I don't know what to add about.

**Comment by just-jeb** [comment]
@muuvmuuv @michelherv I appreciate the effort and thank you both, but currently we have two PRs that do practically the same. Can we please have a single PR that takes the best of the existing PRs?  
Would be much easier for me to follow and comment.  
Thank you!

**Comment by muuvmuuv** [comment]
@michelherv @just-jeb before new year I won't have time for any OSS, so I would close mine and Michel can include all parts you both think that are missing here. 

Below or things I would change or add, doesn't mean they should be included in this PR.

- workspace npm scripts should use binaries directly instead of npx: https://github.com/just-jeb/angular-builders/pull/824/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R15
- puppeteer for karma testing browser bin to avoid using none-existing local binary: https://github.com/just-jeb/angular-builders/pull/824/files#diff-5948cfd5023315c8bda008c2c6688e5a232dd1e0e4932ccfa01db65743069077R3
- use karma-coverage: https://github.com/just-jeb/angular-builders/pull/824/files#diff-5948cfd5023315c8bda008c2c6688e5a232dd1e0e4932ccfa01db65743069077R18
- just use the version without `&copy;` for better test parsing: https://github.com/just-jeb/angular-builders/pull/824/files#diff-f91984e0f1388cbc9fd6c3fb6ba9dbe36b6b08425104d2af479e64ef4b38dfd7R25
- add some note why we use the `node`-typing: https://github.com/just-jeb/angular-builders/pull/824/files#diff-a5e05c6f6ee63c72a42f6f6662bc73589c4c7880a083f414c6441267d53694ebR6
- use file import: https://github.com/just-jeb/angular-builders/pull/824/files#diff-8befd2cce9bc793e8dc30fc050dbbf77f613ef94042f12fb1dca6f1e3b746fc1R8
- fix target option typing bug (appeared in my installation): https://github.com/just-jeb/angular-builders/pull/824/files#diff-b2c00b61854bfb83318b9fa5f3c5110bae9a2fba71f2e9a973059665f6d0651cR17

**Comment by just-jeb** [comment]
Hey, sorry it takes me so long to respond, I'm trying to make the `v11` stable (#908).  
> Hi,
> I updated e2e tests. Can you read the pull request again and tell me if tests are ok for you. Else can you help me by pushing some tests? I will try to improve the solution by TDD.
> 
> About the documentation, I have no improvement idear. `tsConfig` is a `@angular-devkit/architect:browser` property. It's not like `customWebpackConfig` or `indexTransform` which are custom properties. I don't know what to add about.

I'll take a look at tests and leave my comments as a review.  
Regarding documentation; by adding docs I meant that we need to specify in the docs that `indexTransform` can now be written with Typescript, and that both `customWebpackConfig.ts` and `indexTransfrom.ts` are using the app tsConfig now.  
Also if any changes have to be applied to `tsconfig.json` in order to make it work, (like `"types": [ "node" ]`) then we should mention it in the docs as well.

File: .github/workflows/angular-builders.yml [hunk 3]
```diff
@@ -18,12 +18,12 @@ jobs:
         with:
           token: ${{ secrets.GH_PERSONAL_TOKEN }}
 
-      - uses: actions/cache@v1
+      - uses: actions/cache@v2
```
Comment (just-jeb) [inline .github/workflows/angular-builders.yml:21]: I think it's better to keep it as a separate PR, in order to avoid mixing features.

File: packages/custom-webpack/examples/full-cycle-app/angular.json [hunk 4]
```diff
@@ -18,9 +18,12 @@
           "builder": "@angular-builders/custom-webpack:browser",
           "options": {
             "customWebpackConfig": {
-              "path": "./extra-webpack.config.ts"
+              "path": "extra-webpack.config.ts",
+              "mergeStrategies": {
```
Comment (just-jeb) [inline packages/custom-webpack/examples/full-cycle-app/angular.json:22]: [`mergeRules`](https://github.com/just-jeb/angular-builders/blob/master/MIGRATION.MD#migration-from-version-10-to-version-11) should be used instead

File: packages/custom-webpack/examples/full-cycle-app/tsconfig.app.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
   "extends": "./tsconfig.base.json",
   "compilerOptions": {
     "outDir": "./out-tsc/app",
-    "types": []
+    "types": [ "node" ]
```
Comment (just-jeb) [inline packages/custom-webpack/examples/full-cycle-app/tsconfig.app.json:5]: I understand that this is necessary for `ts-node` to work? If yes, maybe we can put it into the hardcoded `compileOptions` when registering the `ts-node`?  
Alternatively the docs should mention that one has to add this to the config when working with `.ts` versions of `customWebpackConfig` or `indexTransform`.

**Review by just-jeb — COMMENTED** [review]
Looks good overall. But I think we also need a test that makes sure `indexTransform` uses the app config as well.

**Comment by michelherv** [comment]
Hi,
I have a problem with the build process
image

Do you understand what is the problem ?

**Comment by just-jeb** [comment]
> Hi,
> I have a problem with the build process
> image
> 
> Do you understand what is the problem ?

Oh, I think I read something about it. As far as I remember it's because it's a fork, and you don't have a token defined in your GH secrets.   
Let me try to find the issue.

**Comment by just-jeb** [comment]
Here it is: https://github.com/actions/checkout/issues/298

**Comment by just-jeb** [comment]
Seems like there is a workaround. I'll try to apply it.

**Comment by just-jeb** [comment]
Aaaand fixed!

**Comment by just-jeb** [comment]
@michelherv are we good to go? Can this be merged?

**Comment by michelherv** [comment]
Hi @just-jeb , it is ok now for merging. Sorry for the late, I was very busy these last days.

According @muuvmuuv [suggestions](https://github.com/just-jeb/angular-builders/pull/879#issuecomment-744495019) :
- I just use the version without &copy; for better test parsing ;
- I add some note why we use the node-typing.

The others suggestions are also very important. I think we should have to treat them as real independent improvements which will be keep anyway independently of this PR. They would can be merge before or after this PR. It it why I opened the PR #925

**Review by just-jeb — APPROVED** [review]

File: packages/custom-webpack/examples/full-cycle-app/angular.json [hunk 6]
```diff
@@ -50,7 +53,6 @@
               "optimization": true,
               "outputHashing": "all",
               "sourceMap": false,
-              "extractCss": true,
```
Comment (tony) [inline packages/custom-webpack/examples/full-cycle-app/angular.json:53]: Any reason why this was removed?

Comment (tony) [reply packages/custom-webpack/examples/full-cycle-app/angular.json:53]: (re: https://github.com/just-jeb/angular-builders/issues/944#issuecomment-863207093)

Comment (tony) [inline packages/custom-webpack/examples/full-cycle-app/angular.json:53]: > Option `"extractCss"` is deprecated: Deprecated since version 11.0. No longer required to disable CSS extraction for HMR.
